### DISCUSSION TITLE
Question: Can this clear big sur notifications?

### DISCUSSION THREAD

**Discussion by torgeir** [babashka/obb#8] [Q&A]
Now if we could make this clear the bloody Big Sur notifications (that they removed keyboard support for) this is a golden addition to my daily emacs setup! 😬 

I've had a few attempts at this some time ago, but to no avail.

Nice work, tho! Feel free to close this at any time.

**Comment by zane** [comment]
> clear the bloody Big Sur notifications

Can you say more about what you mean? You want to use `obb` to dismiss notifications that appear in Notification Center?

**Comment by torgeir** [comment]
Sorry, yes of course.

After upgrading to big sur, the os notifucations have stopped being dismissable by using the keyboard, which is an annoyance to someone like my self that heavily rely on the keyboard. I have been searching for a solution to this problem for some time, and have not found one. It is more a question of scripting big sur in general than related to `obb` I guess - it being possible or not. I just saw an opportunity, as I discovered this nice utility - hoping someone would have a solution at hand. 

What I'd love is some script, e.g. `obb -e '...'`, that I could run on the press of a key to help clear the non-keyboard-navigable os notifications of big sur. These pile up for me constantly, and I end up not looking at them at all.

![Non-dismissed notifications piling up](https://p196.p4.n0.cdn.getcloudapp.com/items/OAu4x4xz/24c0cd9d-a125-4517-b557-ef56125b8bab.jpg?source=viewer&v=a435d1076441fd444f3e63ec1846c46e)

**Comment by zane** [comment]
Would it work to just uncheck "Show in Notification Center" for the offending apps in the "Notifications & Focus" preference pane? (Names are slightly different in Big Sur; I'm on Monterey.)

**Comment by torgeir** [comment]
To remove them completely, yes. But I’d rather be able to clear when Ive seen them and actually make use of them. 

Are they keyboard navigable in Monterey? They were before I upgraded to big sur..

**Reply by zane** [reply]
If they are I'm not sure how. Keyboard preference pane > Shortcuts > Mission Control shows only an option to bind a key to show the Notification Center.

**Comment by zane** [comment]
I think what you want should be possible via [UI scripting](https://github.com/JXA-Cookbook/JXA-Cookbook/wiki/System-Events#clicking-menu-items). 

This should get you started:

``` clojure
(def system-events (js/Application "System Events"))

(defn application-process
  [s]
  (-> system-events
    (.-applicationProcesses)
    (.byName s)))

(defn process
  [s]
  (-> system-events
      (.-processes)
      (.byName s)))

;; Open the control center by clicking on the time in the menu bar.
(-> (application-process "ControlCenter")
    (.-menuBars)
    (.at 0)
    (.-menuBarItems)
    (.byName "Clock")
    (.click))

;; Find and click the close button.
(-> (process "Notification Center")
    (.-windows)
    (.byName "Notification Center")
    ;; FIXME
    )
```
In order to find and click the close button you'll probably want to [use the accessibility inspector](https://developer.apple.com/library/archive/documentation/Accessibility/Conceptual/AccessibilityMacOSX/OSXAXTestingApps.html).

**Reply by zane** [reply]
You'll also need to change the above slightly to make it work in Big Sur. At minimum you'll need to click on a different menu bar item to open the control center.
